cirq/google/calibration/engine_simulator.py

```diff
@@ -0,0 +1,405 @@
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+    cast,
+)
+
+import numpy as np
+
+from cirq.circuits import Circuit, PointOptimizer, PointOptimizationSummary
+from cirq.ops import (
+    FSimGate,
+    Gate,
+    MeasurementGate,
+    Operation,
+    PhasedFSimGate,
+    Qid,
+    QubitOrderOrList,
+    SingleQubitGate,
+    WaitGate,
+)
+from cirq.sim import (
+    Simulator,
+    SimulatesSamples,
+    SimulatesIntermediateStateVector,
+    SparseSimulatorStep,
+    StepResult,
+)
+from cirq.study import ParamResolver
+from cirq.value import RANDOM_STATE_OR_SEED_LIKE, parse_random_state
+
+from cirq.google.calibration.phased_fsim import (
+    FloquetPhasedFSimCalibrationRequest,
+    IncompatibleMomentError,
+    PhasedFSimCalibrationRequest,
+    PhasedFSimCalibrationResult,
+    PhasedFSimCharacterization,
+    SQRT_ISWAP_PARAMETERS,
+    try_convert_sqrt_iswap_to_fsim,
+)
+
+
+ParametersDriftGenerator = Callable[[Qid, Qid, FSimGate], PhasedFSimCharacterization]
+PhasedFsimDictParameters = Dict[
+    Tuple[Qid, Qid], Union[Dict[str, float], PhasedFSimCharacterization]
+]
+
+
+class PhasedFSimEngineSimulator(SimulatesSamples, SimulatesIntermediateStateVector):
+    """Wrapper on top of cirq.Simulator that allows to simulate calibration requests.
+
+    This simulator introduces get_calibrations which allows to simulate
+    cirq.google.run_characterizations requests. The returned calibration results represent the
+    internal state of a simulator. Circuits which are run on this simulator are modified to account
+    for the changes in the unitary parameters as described by the calibration results.
+    """
+
+    def __init__(
+        self,
+        simulator: Simulator,
+        *,
+        drift_generator: ParametersDriftGenerator,
+        gates_translator: Callable[[Gate], Optional[FSimGate]] = try_convert_sqrt_iswap_to_fsim,
```

Comment:
why are the gates all `FSimGate` instead of `PhasedFSimGate`?

Reply:
I guess it was not really important to me. I wanted to connect it more to the hardware gates we plan to support: it's very unlikely that we'd like to have `PhasedFSimGate` supported on the engine API level so `FSimGate` describes better what's applicable. I can change it, although it will trigger changes beyond this PR.